### PR TITLE
Transaction fees oracle

### PR DESCRIPTION
**Pull request by bingen**
Replaces #270

### CODE REVIEW

**Comment by facuspagnuolo** [comment]
@bingen changed the target branch from `master` to `development`

**Comment by facuspagnuolo** [comment]
@bingen would you mind changing the target branch from `master` to `development`

File: test/subscriptions/aragon-app-fees-cashier.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+const { bigExp } = require('../helpers/lib/numbers')
+const { TRANSACTIONS_FEES_ORACLE_ERRORS, CONTROLLED_ERRORS } = require('../helpers/utils/errors')
+
+const AragonAppFeesCashier = artifacts.require('AragonAppFeesCashierMock')
+const ERC20 = artifacts.require('ERC20Mock')
+
+const VOTING_APP_ID = '0x9fa3927f639745e587912d4b0fea7ef9013bf93fb907d29faeab57417ba6e1d4'
+const TOKEN_MANAGER_APP_ID = '0x6b20a3010614eeebf2138ccec99f028a61c811b3b1a3343b6ff635985c75c91f'
+
+contract('Transaction Fees Oracle', ([_, governor, subscriber, fakeToken]) => {
```
Comment (facuspagnuolo) [inline test/subscriptions/aragon-app-fees-cashier.js:13]:
```suggestion
contract('Aragon App Fees Cashier', ([_, governor, subscriber, fakeToken]) => {
```

File: contracts/lib/os/EtherTokenConstant.sol [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/*
+ * SPDX-License-Identifier:    MIT
+ */
+
+pragma solidity ^0.5.8;
+
+
+// subscription fees oracle rely on address(0) to denote native ETH
```
Comment (facuspagnuolo) [inline contracts/lib/os/EtherTokenConstant.sol:8]:
```suggestion
// app fees cashier rely on address(0) to denote native ETH
```

File: contracts/subscriptions/AragonAppFeesCashier.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    struct AppFee {
+        bool set;
+        ERC20 token;
+        uint256 amount;
+    }
+
+    mapping (bytes32 => AppFee) internal appFees;
+
+    /**
+    * @dev Initialize court subscription fees oracle
```
Comment (facuspagnuolo) [inline contracts/subscriptions/AragonAppFeesCashier.sol:26]:
```suggestion
    * @dev Initialize Aragon app fees cashier
```

File: test/helpers/utils/errors.js [hunk 4]
```diff
@@ -89,6 +89,14 @@ const SUBSCRIPTIONS_ERRORS = {
   SUBSCRIPTION_NOT_PAUSED: 'CS_SUBSCRIPTION_NOT_PAUSED'
 }
 
+const TRANSACTIONS_FEES_ORACLE_ERRORS = {
+  ERROR_APP_NOT_SET: 'SFO_APP_NOT_SET',
+  ERROR_WRONG_TOKEN: 'SFO_WRONG_TOKEN',
+  ERROR_COURT_HAS_NOT_STARTED: 'SFO_COURT_HAS_NOT_STARTED',
+  ERROR_WRONG_TOKENS_LENGTH: 'SFO_WRONG_TOKENS_LENGTH',
+  ERROR_WRONG_AMOUNTS_LENGTH: 'SFO_WRONG_AMOUNTS_LENGTH'
+}
```
Comment (facuspagnuolo) [inline test/helpers/utils/errors.js:98]: Let's rename these properly

File: contracts/subscriptions/AragonAppFeesCashier.sol [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    */
+    function unsetAppFees(bytes32[] calldata _appIds) external onlyConfigGovernor {
+        for (uint256 i = 0; i < _appIds.length; i++) {
+            _unsetAppFee(_appIds[i]);
+        }
+    }
+
+    // TODO: To be integrated with CourtSubscriptions with the new trusted model
+    function payAppFees(bytes32 _appId, bytes calldata _data) external {
+        emit AppFeePaid(msg.sender, _appId, _data);
```
Comment (facuspagnuolo) [inline contracts/subscriptions/AragonAppFeesCashier.sol:78]: We can probably use the same pattern we followed for other modules here. Basically introduce an internal method that must be implemented by the sub-contract to make sure you don't miss it. In this case I'd something like:

```suggestion
        AppFee storage appFee = appFees[_appId];
        require(appFee.set, ERROR_APP_ID_NOT_SET);
        
        _receiveAppFees(msg.sender, _appId, appFee.token, appFee.amount, _data);
        emit AppFeePaid(msg.sender, _appId, _data);
```

File: contracts/subscriptions/AragonAppFeesCashier.sol [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+import "./IAragonAppFeesCashier.sol";
+
+
+contract AragonAppFeesCashier is Controlled, IAragonAppFeesCashier, EtherTokenConstant {
+    string private constant ERROR_APP_ID_NOT_SET = "SFO_APP_NOT_SET";
+    string private constant ERROR_WRONG_TOKEN = "SFO_WRONG_TOKEN";
+    string private constant ERROR_COURT_HAS_NOT_STARTED = "SFO_COURT_HAS_NOT_STARTED";
+    string private constant ERROR_WRONG_TOKENS_LENGTH = "SFO_WRONG_TOKENS_LENGTH";
+    string private constant ERROR_WRONG_AMOUNTS_LENGTH = "SFO_WRONG_AMOUNTS_LENGTH";
```
Comment (facuspagnuolo) [inline contracts/subscriptions/AragonAppFeesCashier.sol:15]: Let's rename these

**Comment by bingen** [comment]
> We can probably merge the trusted model PR first so we can review the integration properly

Yes, I agree, but I did it the other way around, pushing soon.

File: contracts/subscriptions/AragonAppFeesCashier.sol (hunk 5 above)
Comment (bingen) [reply contracts/subscriptions/AragonAppFeesCashier.sol:78]: Yes, but as I noted in the header, I left this as a TODO for the integration ;-)

**Pull request merged**